### PR TITLE
Better model panics

### PR DESCRIPTION
**Pull request by IamTheCarl**
Mostly ~resolves~ addresses https://github.com/hannobraun/Fornjot/issues/1569 . (edited by @hannobraun: "resolves" is a magic word that closes the issue; GitHub will readily take instructions from random words within the pull request description, but doesn't know what "mostly" means)

It currently does not print the stack trace, but it does print the fact that it's a panic and a line number where the panic occurred. Given a little more time I might be able to get the stack trace.

image

The key is to override the panic hook within the model. This cannot be done externally, so the initialization code is injected into the model function by the `#[fj::model]` macro. That new panic hook then places all panics into a global variable for the `on_panic` function to grab. I suspect this could get a bit funky when multiple threads are involved, but the worst I would expect is it giving a "no details provided" message or the wrong error message on a panic.

### CODE REVIEW

File: crates/fj-proc/src/lib.rs [hunk 1]
```diff
@@ -91,13 +95,67 @@ pub fn model(_: TokenStream, input: TokenStream) -> TokenStream {
...
+                            },
+                        })),
+                        paren_token: Paren {
+                            span: Span::call_site(),
+                        },
+                        args: Punctuated::new(),
+                    }),
+                    syn::token::Semi::default(),
+                ),
+            );
```
Comment (hannobraun) [inline crates/fj-proc/src/lib.rs:149]: I'm far from an expert in procedural macros and might be wrong about what's happening here, but couldn't this be done much easier using the `quote!` macro? See [`expand.rs`](https://github.com/hannobraun/Fornjot/blob/main/crates/fj-proc/src/expand.rs) for examples.

File: crates/fj-proc/src/lib.rs [hunk 2]
```diff
@@ -91,13 +95,67 @@ pub fn model(_: TokenStream, input: TokenStream) -> TokenStream {
...
 
             let tokens = quote::quote! {
                 #item
                 #init
+
             };
 
+            eprintln!("TOKENS: {}", tokens);
+
```
Comment (hannobraun) [inline crates/fj-proc/src/lib.rs:158]: It it intentional that this was left in here? Or just leftover debug output?

**Review by hannobraun — APPROVED** [review]
Thank you, @IamTheCarl! Looks good and works great.

I've left two comments with suggestions for potential improvements. None of those are critical, so I'm going to merge this now. But feel free to send a follow-up PR, if you agree with them.

I also made a small edit to your pull request description, to prevent #1569 from being closed when I merge this.

File: crates/fj-proc/src/lib.rs (hunk 2 above)
Comment (IamTheCarl) [reply crates/fj-proc/src/lib.rs:158]: Leftover debug output... I thought I got them all...

**Pull request merged**